contributors/design-proposals/testing/trigger-pr-tests.md

```diff
@@ -0,0 +1,92 @@
+# Triggering PR tests
+
+## Objective
+
+This document describes a major change to the way the `trigger` plugin determines if test jobs should be started on a pull request (PR).
+
+## Current situation
+
+PR test jobs are started by the trigger plugin on **trusted PR** events, or when a untrusted PR becomes trusted.
+> A PR is considered trusted if the author is a member of the *trusted organization* for the repository or if such a member has left an `/ok-to-test` command on the PR.
+
+You can easily spot an untrusted PR by its `needs-ok-to-test` label.
+
+## Problems
+
+### Reviews are not parsed by the trigger plugin
+
+We only parse PR comments for the `/ok-to-test` command, not review comments.
+
+@spiffxp
+> I still see people using /ok-to-test in PR reviews (as comments, aka no approval or change requested) and nothing happening
+
+@cjwagner
+> We aren't currently fetching the reviews, just the comments.
+
+### API tokens consumption
+
+@cjwagner
+> The `trigger` plugin has to list all comments to see if an `/ok-to-test` command exists every time it wants to check that a non-member PR is trusted. Checking reviews and review comments for this command would require at least 2 additional API tokens every time.
+
+> Fetching the reviews and review comments would use a lot more API tokens. If we switch to a positive label we don't need to check anything except the labels though.
+
+### Code complexity
+
+The `trigger` code is already quite complex, and also parsing the reviews would make it worse.
+
+@cjwagner
+> We could check `review_comment` events and remove the `do-not-merge/needs-ok-to-test` label if the comment contains `/ok-to-test`, but we would still need to list all reviews and review comments when checking if a non-member PR is trusted to prevent race conditions with other events.
+
+> Using a positive label instead of a negative one avoids all the race conditions that force us to list comments and check for `/ok-to-test` instead of just checking for a label.
+
+## Proposal
+
+We suggest introducing a new label named `ok-to-test` that would be required on any non-member PR before automatic test jobs can be started by the `trigger` plugin.
+
+This label could only be added by members of the *trusted organization* for the repository.
+
+We will also deprecate the `/ok-to-test` command.
+
+## Caveats
+
+Switching from a negative to a positive label could create confusion and add clutter to the PR.
+
+@cblecker
+> .. label overload. Having a label just to say testing is okay.. bleh.
```

Comment:
Could we make this a github context? This would:
* allow us to only flag PRs that are unsafe for testing
* accept ok-to-test from either an issue or review comment
* simplify searching (just look for the contexts, rather than all comments

Reply:
other pros/cons:

\+ A github context would also not be human editable which avoids accidentally removing it when editing labels

\- A context is not searchable by humans AFAIK, right now we have [go.k8s.io/needs-ok-to-test](https://go.k8s.io/needs-ok-to-test) for contribex to help onboard new people, if it's still a label we can update this github search link to still work..

\+/- github contexts are cleared when a PR is closed/reopened

\- our current workflow mostly involves labels and commands that manipulate them, should approval / lgtm also be a context..?

Reply:
Ok then this is a no show. I will add this option in the doc and explain why it's not a good idea.